src/mame/machine/dmac3.cpp

```diff
@@ -0,0 +1,231 @@
+	if (data & CSR_RESET)
+	{
+		LOGMASKED(LOG_GENERAL, "dmac3-%d chip reset\n", controller);
+		reset_controller(controller);
+	}
+	else
+	{
+		m_controllers[controller].csr = data;
+	}
+}
+
+void dmac3_device::intr_w(dmac3_controller controller, uint32_t data)
+{
+	static constexpr uint32_t INTR_CLR_MASK = (INTR_INT | INTR_TCI | INTR_EOP | INTR_EOPI | INTR_DREQ | INTR_DRQI | INTR_PERR);
+	static constexpr uint32_t INTR_EN_MASK = (INTR_INTEN | INTR_TCIE | INTR_EOPIE | INTR_DRQIE);
+
+	LOGMASKED(LOG_REGISTER, "dmac3-%d intr_w: 0x%x\n", controller, data);
+	const auto intr_clear_bits = data & INTR_CLR_MASK;  // Get 1s on bits to clear
+	const auto intr_enable_bits = data & INTR_EN_MASK;  // Get 1s on bits to set
+	m_controllers[controller].intr &= ~intr_clear_bits; // Clear requested interrupt flags
+	m_controllers[controller].intr &= ~INTR_EN_MASK;    // Clear all mask bits
+	m_controllers[controller].intr |= intr_enable_bits; // Set mask bits to new mask
+	m_irq_check->adjust(attotime::zero);
+}
+
+void dmac3_device::length_w(dmac3_controller controller, uint32_t data)
+{
+	LOGMASKED(LOG_REGISTER, "dmac3-%d length_w: 0x%x (%s)\n", controller, data, machine().describe_context());
+	m_controllers[controller].length = data;
+}
+
+void dmac3_device::address_w(dmac3_controller controller, uint32_t data)
+{
+	LOGMASKED(LOG_REGISTER, "dmac3-%d address_w: 0x%x (%s)\n", controller, data, machine().describe_context());
+	m_controllers[controller].address = data;
+}
+
+void dmac3_device::conf_w(dmac3_controller controller, uint32_t data)
+{
+#if (VERBOSE & LOG_REGISTER) > 0 
+	// No need for the extra comparison if logging isn't enabled
+	// Only log if something other than the access mode changed
+	// since, at least for now, DMAC and SPIFI accesses will go through regardless
+	// of this setting.
+
+	if ((data & ~CONF_WIDTH) != (m_controllers[controller].conf & ~CONF_WIDTH))
+	{
+		LOGMASKED(LOG_REGISTER, "dmac3-%d conf_w: 0x%x (%s)\n", controller, data, machine().describe_context());
+	}
+#endif
+
+	m_controllers[controller].conf = data;
+}
+
+void dmac3_device::device_reset()
+{
+	// Reset both controllers
+	reset_controller(dmac3_controller::CTRL0);
+	reset_controller(dmac3_controller::CTRL1);
+}
+
+void dmac3_device::reset_controller(dmac3_controller controller)
+{
+	m_controllers[controller].csr = 0;
+	m_controllers[controller].intr &= INTR_INT; // TODO: is the external interrupt bit preserved?
+	m_controllers[controller].length = 0;
+	m_controllers[controller].address = 0;
+	m_controllers[controller].conf = 0;
+	m_irq_check->adjust(attotime::zero);
+}
+
+TIMER_CALLBACK_MEMBER(dmac3_device::irq_check)
+{
+	bool newIrq = false;
+
+	// Scan each controller for an interrupt condition - if any of these are true, set IRQ.
+	// If both controllers have no interrupt conditions, IRQ can be cleared.
+	for (int controller = 0; controller < 2; ++controller)
+	{
+		const uint32_t intr = m_controllers[controller].intr;
```

Comment:
You could do this with a range-based for loop, like:
```
    for (const auto &controller : m_controllers)
    {
        const uint32_t intr = controller.intr;
        ...
```

It saves having to put the size of the array here, and it’s a bit easier on the eyes when you aren’t using the index for anything other than indexing the array.